lightning-invoice/src/utils.rs

```diff
@@ -332,6 +332,82 @@ where
 	)
 }
 
+pub fn create_invoice_from_channelmanager_and_duration_since_epoch_with_payment_hash<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>(
+	channelmanager: &ChannelManager<M, T, K, F, L>, keys_manager: K, logger: L,
+	network: Currency, amt_msat: Option<u64>, description: String, duration_since_epoch: Duration,
+	invoice_expiry_delta_secs: u32, payment_hash: PaymentHash, payment_secret: PaymentSecret
+) -> Result<Invoice, SignOrCreationError<()>>
+	where
+		M::Target: chain::Watch<<K::Target as KeysInterface>::Signer>,
+		T::Target: BroadcasterInterface,
+		K::Target: KeysInterface,
+		F::Target: FeeEstimator,
+		L::Target: Logger,
+{
+	_create_invoice_from_channelmanager_and_duration_since_epoch_with_payment_hash(
+		channelmanager, keys_manager, logger, network, amt_msat,
+		InvoiceDescription::Direct(
+			&Description::new(description).map_err(SignOrCreationError::CreationError)?,
+		),
+		duration_since_epoch, invoice_expiry_delta_secs, payment_hash, payment_secret
+	)
+}
+
+fn _create_invoice_from_channelmanager_and_duration_since_epoch_with_payment_hash<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>(
```

Comment:
+1 for DRYing this is up.

Reply:
@TheBlueMatt , do you mean change ``_create_invoice_from_channelmanager_and_duration_since_epoch`` so that it also takes a payment_hash and have it also construct the payment_secret inside (as suggested in your other comment). That way, my main function calls this method instead of my new one?

Reply:
thank you! I hope this new commit fixes this as envisioned and addresses the other comments here.